encoding/thrift/inbound_nowire.go

```diff
@@ -0,0 +1,155 @@
+}
+
+type NoWireCall struct {
+	Reader        io.Reader
+	RequestReader stream.RequestReader
+	EnvelopeType  wire.EnvelopeType
+
+	StreamReader stream.Reader
+}
+
+type NoWireHandler interface {
+	Handle(context.Context, *NoWireCall) (NoWireResponse, error)
+}
+
+func (t thriftNoWireHandler) Handle(ctx context.Context, treq *transport.Request, rw transport.ResponseWriter) error {
+	ctx, call := encodingapi.NewInboundCall(ctx)
+	res, err := doRequest(ctx, call, treq, rw, t.NoWireHandler, wire.Call, t.Protocol, t.Enveloping)
+	if err != nil {
+	}
+
+	if resType := res.Body.EnvelopeType(); resType != wire.Reply {
+		return errors.ResponseBodyEncodeError(
+			treq, errUnexpectedEnvelopeType(resType))
+	}
+
+	if res.IsApplicationError {
+		rw.SetApplicationError()
+
+		if applicationErrorMetaSetter, ok := rw.(transport.ApplicationErrorMetaSetter); ok {
+			applicationErrorMetaSetter.SetApplicationErrorMeta(&transport.ApplicationErrorMeta{
+				Details: res.ApplicationErrorDetails,
+				Name:    res.ApplicationErrorName,
+				Code:    res.ApplicationErrorCode,
+			})
+		}
+	}
+
+	if err := call.WriteToResponse(rw); err != nil {
+		// not reachable
+		return err
+	}
+
+	if err := res.ResponseWriter.WriteResponse(wire.Reply, rw, res.Body); err != nil {
+		return errors.ResponseBodyEncodeError(treq, err)
+	}
+
+	return nil
+}
+
+func (t thriftNoWireHandler) HandleOneway(ctx context.Context, treq *transport.Request) error {
+	ctx, call := encodingapi.NewInboundCall(ctx)
+	_, err := doRequest(ctx, call, treq, nil, t.NoWireHandler, wire.OneWay, t.Protocol, t.Enveloping)
+	return err
+}
+
+func doRequest(
+	ctx context.Context,
+	call *encodingapi.InboundCall,
+	treq *transport.Request,
+	rw transport.ResponseWriter,
+	noWireHandler NoWireHandler,
+	reqEnvelopeType wire.EnvelopeType,
+	proto stream.Protocol,
+	enveloping bool,
+) (NoWireResponse, error) {
+	if err := errors.ExpectEncodings(treq, Encoding); err != nil {
+		return _emptyResponse, err
+	}
+
+	if err := call.ReadFromRequest(treq); err != nil {
+		return _emptyResponse, err
+	}
+
+	if reqReader, ok := proto.(stream.RequestReader); ok {
+		nwc := NoWireCall{
+			Reader:        treq.Body,
+			RequestReader: reqReader,
+			EnvelopeType:  reqEnvelopeType,
+		}
+		return noWireHandler.Handle(ctx, &nwc)
+	}
+
+	streamReader := binary.NewStreamReader(treq.Body)
+	defer streamReader.Close()
+
+	if enveloping {
+		eh, err := streamReader.ReadEnvelopeBegin()
+		if err != nil {
+			return _emptyResponse, err
+		}
+		if eh.Type != reqEnvelopeType {
+			return _emptyResponse, errors.RequestBodyDecodeError(treq, errUnexpectedEnvelopeType(eh.Type))
+		}
+
+		responder := binary.EnvelopeV1Responder{Name: eh.Name, SeqID: eh.SeqID}
+		nwc := NoWireCall{StreamReader: streamReader}
+		resp, err := noWireHandler.Handle(ctx, &nwc)
+		if err != nil {
+			return _emptyResponse, err
+		}
+
+		resp.ResponseWriter = responder
+		return resp, streamReader.ReadEnvelopeEnd()
+	}
+
+	responder := binary.NoEnvelopeResponder
```

Comment:
move closer to where it is used L153 and inline due to single use. Same suggestion to L135
`resp.ResponseWriter = binary.NoEnvelopeResponder`